### PR TITLE
Document Kubeflow integration with lakeFS

### LINKED ISSUE #2093: Docs: Using Kubeflow with lakefs
Document the current way to use Kubeflow with lakefs. 

The approach is to define a container operation built with a lakectl image and create pipeline tasks that run any lakectl command. similarly to what we did with Airflow before publishing our own operators. (https://lakefs.io/building-reproducible-data-pipelines-with-airflow-and-lakefs/)

### PR DESCRIPTION
**Pull request by talSofer**
closes #2093

### CODE REVIEW

File: docs/_config.yml [hunk 1]
```diff
@@ -67,4 +67,4 @@ plugins:
   - jekyll-redirect-from
   - jekyll-seo-tag
 
-exclude: ["deploy/includes"]
+exclude: ["deploy/includes", "vendor"]
```
Comment (johnnyaug) [inline docs/_config.yml:70]: What is this?

File: docs/integrations/kubeflow.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+A Kubeflow pipeline is a portable and scalable definition of an ML workflow composed of steps. Each step in the pipeline is an instance of a component represented as an instance of [ContainerOp](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ContainerOp).
+
+There are some aspects we will need to handle to run Kubeflow with lakeFS:
+## Install lakeFS CLI 
+lakeFS comes with its own native CLI client. We will use it to build lakeFS-specific Kubeflow pipeline steps. 
+To install and configure the lakeFS CLI client follow [this](https://docs.lakefs.io/quickstart/lakefs_cli.html) guide.  
+
+## Add lakeFS-specific pipeline steps
+Once you have `lakectl` installed, you can add pipeline components that invoke lakeFS operations. As mentioned above, a pipeline component is represented as a ContainerOp.
+For `lakectl` to work with Kubeflow, you will need to provide your `lakectl` configurations as environment variables named:
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:31]: Do I need those even if I follow the documentation on how to configure lakectl, using for example `lakectl config`? The relation between this paragraph and the previous one (titled "Install lakeFS CLI") is not clear.

File: docs/integrations/kubeflow.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+* `LAKECTL_CREDENTIALS_ACCESS_KEY_ID: <Access key ID>`
+* `LAKEFS_SECRET_ACCESS_KEY: <Secret access key>`
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:33]: the first one is `LAKECTL_` and the second one is `LAKEFS_`? :thinking:

File: docs/integrations/kubeflow.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+* `LAKECTL_SERVER_ENDPOINT_URL: <Server endpoint URL>`
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:34]: Better to use examples for these values rather than placeholders, IMO.
e.g.:
`AKIAIOSFODNN7EXAMPLE` 
`https://lakefs.example.com/api/v1/`

File: docs/integrations/kubeflow.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    
+   ```python
+    def merge_op():
+        return dsl.ContainerOp(
+           name='merge',
+           image='treeverse/lakectl',
+           arguments=['merge', 'lakefs://repo/example-branch', 'lakefs://example-repo/main']).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_ACCESS_KEY_ID',value='<Access key ID>')).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_SECRET_ACCESS_KEY',value='<Secret access key>')).add_env_variable(V1EnvVar(name='LAKECTL_SERVER_ENDPOINT_URL',value='<Server endpoint URL>'))
+    ```
+
+You can invoke any lakeFS operation supported by `lakectl` by implementing it a ContainerOp. Check out the complete [CLI reference](https://docs.lakefs.io/reference/commands.html) for the list of supported operations.
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:70]:
```suggestion
You can invoke any lakeFS operation supported by `lakectl` by implementing it as a ContainerOp. Check out the complete [CLI reference](https://docs.lakefs.io/reference/commands.html) for the list of supported operations.  
```

File: docs/integrations/kubeflow.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+## Add the lakeFS steps to your pipeline
+Add the steps created on the previous step to your pipeline before compiling it. 
+
+### Example pipeline
+{: .no_toc }
+A pipeline that implements a simple ETL, that has steps for branch creation and commits.    
+   ```python
+    def lakectl_pipeline():
+       create_branch_task = create_branch_op()
+       extract_task = extract_op() # a made-up generic operation
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:81]:
```suggestion
       extract_task = example_extract_op()
```

File: docs/integrations/kubeflow.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+### Example pipeline
+{: .no_toc }
+A pipeline that implements a simple ETL, that has steps for branch creation and commits.    
+   ```python
+    def lakectl_pipeline():
+       create_branch_task = create_branch_op()
+       extract_task = extract_op() # a made-up generic operation 
+       commit_task = commit_op() 
+       transform_task = transform_op() # a made-up generic operation
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:83]:
```suggestion
       transform_task = example_transform_op()
```

File: docs/integrations/kubeflow.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+   ```
+
+## What's next?
+{: .no_toc }
+We are working to make the lakeFS [python client](https://docs.lakefs.io/integrations/python.html) work with Kubeflow.
+
+
+
+**Notes**  \  
+lakeFS version <= v0.33.1 uses '@' (instead of '/') as separator between repository and branch.  \
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:95]: Is this relevant here?

File: docs/integrations/kubeflow.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+1. TOC 
+{:toc}
+
+## Using lakeFS with Kubeflow pipelines
+{: .no_toc }
+A Kubeflow pipeline is a portable and scalable definition of an ML workflow composed of steps. Each step in the pipeline is an instance of a component represented as an instance of [ContainerOp](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ContainerOp).
+
+There are some aspects we will need to handle to run Kubeflow with lakeFS:
+## Install lakeFS CLI
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:25]:
```suggestion
## Prerequisites
### Install lakeFS CLI 
```

**Review by johnnyaug — CHANGES_REQUESTED** [review]

File: docs/integrations/kubeflow.md (hunk 4 above)
Comment (nopcoder) [reply docs/integrations/kubeflow.md:34]: even without the '/api/v1/' part - the lakectl will the use current base uri

File: docs/integrations/kubeflow.md (hunk 3 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:33]: no both are lakectl, good catch thx :)

File: docs/integrations/kubeflow.md (hunk 8 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:95]: yes because the operation includes branches and repo paths. WDYT?

Comment (johnnyaug) [reply docs/integrations/kubeflow.md:95]: But you only support v>0.43 anyways

File: docs/integrations/kubeflow.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+## What's next?
+{: .no_toc }
+We are working to make the lakeFS [python client](https://docs.lakefs.io/integrations/python.html) work with Kubeflow.
+
+
+
+**Notes**  \  
+lakeFS version <= v0.33.1 uses '@' (instead of '/') as separator between repository and branch.  \
+The lakeFS Kubeflow integration is supported on lakeFS version >= v0.43.0.
+{: .note}
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:97]:
```suggestion
{: .note }
```

File: docs/integrations/kubeflow.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+       extract_task = extract_op() # a made-up generic operation 
+       commit_task = commit_op() 
+       transform_task = transform_op() # a made-up generic operation
+       commit_task = commit_op()
+       load_task = load_op() # a made-up generic operation
+   ```
+
+## What's next?
+{: .no_toc }
+We are working to make the lakeFS [python client](https://docs.lakefs.io/integrations/python.html) work with Kubeflow.
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:90]: not sure we want this one here - it is hard to follow what we are working on if it is not part of the roadmap or issues.

File: docs/integrations/kubeflow.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+
+
+**Notes**  \
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:94]:
```suggestion
**Note**
```

File: docs/integrations/kubeflow.md (hunk 11 above)
Comment (nopcoder) [reply docs/integrations/kubeflow.md:90]: don't reference the full path in the doc to the docs. you want the relative path to the markdown file. usually, full path means it is an external link and also usually open in a different tab.

File: docs/integrations/kubeflow.md (hunk 9 above)
Comment (nopcoder) [reply docs/integrations/kubeflow.md:25]: Recommend to use a blank line between headers and content - or any markdown block.

File: docs/integrations/kubeflow.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+{:toc}
+
+## Using lakeFS with Kubeflow pipelines
+{: .no_toc }
+A Kubeflow pipeline is a portable and scalable definition of an ML workflow composed of steps. Each step in the pipeline is an instance of a component represented as an instance of [ContainerOp](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ContainerOp).
+
+There are some aspects we will need to handle to run Kubeflow with lakeFS:
+## Install lakeFS CLI 
+lakeFS comes with its own native CLI client. We will use it to build lakeFS-specific Kubeflow pipeline steps. 
+To install and configure the lakeFS CLI client follow [this](https://docs.lakefs.io/quickstart/lakefs_cli.html) guide.
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:27]: 1. relative link to markdown
2. wrap 'this guide' as a link

File: docs/integrations/kubeflow.md (hunk 2 above)
Comment (nopcoder) [reply docs/integrations/kubeflow.md:31]: In this example, we use env variables to pass the lakefs configuration.
1. the end-user can use volume and mount our lakectl yaml configuration (but we don't need to explain everything) it is not just a must
2. we should look into how to pass secrets to the operation

File: docs/integrations/kubeflow.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    from kfp import dsl
+    from kubernetes.client.models import V1EnvVar
+    
+    def create_branch_op():
+       return dsl.ContainerOp(
+          name='create_branch',
+          image='treeverse/lakectl',
+          arguments=['branch', 'create', 'lakefs://example-repo/example-branch', '-s', 'lakefs://example-repo/main']).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_ACCESS_KEY_ID',value='<Access key ID>')).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_SECRET_ACCESS_KEY',value='<Secret access key>')).add_env_variable(V1EnvVar(name='LAKECTL_SERVER_ENDPOINT_URL',value='<Server endpoint URL>'))
+    ```
+2. Commit changes to a branch
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:51]:
```suggestion
    ```

2. Commit changes to a branch
```

File: docs/integrations/kubeflow.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+   ```python
+    import kfp
+    from kfp import dsl
+    from kubernetes.client.models import V1EnvVar
+    
+    def create_branch_op():
+       return dsl.ContainerOp(
+          name='create_branch',
+          image='treeverse/lakectl',
+          arguments=['branch', 'create', 'lakefs://example-repo/example-branch', '-s', 'lakefs://example-repo/main']).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_ACCESS_KEY_ID',value='<Access key ID>')).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_SECRET_ACCESS_KEY',value='<Secret access key>')).add_env_variable(V1EnvVar(name='LAKECTL_SERVER_ENDPOINT_URL',value='<Server endpoint URL>'))
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:49]: 1. can you align this line to have add_env_varialbe one under the other?
2. looks like calling the add_env.. for each operation can be extracted to a function that we can call for each example. or at least will capture what we are doing for the user - like `add_lakectl_env`

File: docs/integrations/kubeflow.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+## Add lakeFS-specific pipeline steps
+Once you have `lakectl` installed, you can add pipeline components that invoke lakeFS operations. As mentioned above, a pipeline component is represented as a ContainerOp.
+For `lakectl` to work with Kubeflow, you will need to provide your `lakectl` configurations as environment variables named:
+* `LAKECTL_CREDENTIALS_ACCESS_KEY_ID: <Access key ID>`
+* `LAKEFS_SECRET_ACCESS_KEY: <Secret access key>`
+* `LAKECTL_SERVER_ENDPOINT_URL: <Server endpoint URL>`
+
+### Example operations
+{: .no_toc }
+1. Create new branch
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:38]: maybe add a note that we crate a branch called 'example-branch' based on 'master' for 'example-repo' repository.

File: docs/integrations/kubeflow.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    ```
+3. Merge two lakeFS branches
+    
+   ```python
+    def merge_op():
+        return dsl.ContainerOp(
+           name='merge',
+           image='treeverse/lakectl',
+           arguments=['merge', 'lakefs://repo/example-branch', 'lakefs://example-repo/main']).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_ACCESS_KEY_ID',value='<Access key ID>')).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_SECRET_ACCESS_KEY',value='<Secret access key>')).add_env_variable(V1EnvVar(name='LAKECTL_SERVER_ENDPOINT_URL',value='<Server endpoint URL>'))
+    ```
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:68]: your code is one space inside the block

File: docs/integrations/kubeflow.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+* `LAKECTL_CREDENTIALS_ACCESS_KEY_ID: <Access key ID>`
+* `LAKEFS_SECRET_ACCESS_KEY: <Secret access key>`
+* `LAKECTL_SERVER_ENDPOINT_URL: <Server endpoint URL>`
+
+### Example operations
+{: .no_toc }
+1. Create new branch 
+    
+   ```python
+    import kfp
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:41]: this block code is one space inside (you will get a space in the start of each line)

File: docs/integrations/kubeflow.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+           arguments=['merge', 'lakefs://repo/example-branch', 'lakefs://example-repo/main']).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_ACCESS_KEY_ID',value='<Access key ID>')).add_env_variable(V1EnvVar(name='LAKECTL_CREDENTIALS_SECRET_ACCESS_KEY',value='<Secret access key>')).add_env_variable(V1EnvVar(name='LAKECTL_SERVER_ENDPOINT_URL',value='<Server endpoint URL>'))
+    ```
+
+You can invoke any lakeFS operation supported by `lakectl` by implementing it a ContainerOp. Check out the complete [CLI reference](https://docs.lakefs.io/reference/commands.html) for the list of supported operations.  
+
+## Add the lakeFS steps to your pipeline
+Add the steps created on the previous step to your pipeline before compiling it. 
+
+### Example pipeline
+{: .no_toc }
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:76]: do you need this one also here? think it is required once for the top level header

File: docs/integrations/kubeflow.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+A pipeline that implements a simple ETL, that has steps for branch creation and commits.    
+   ```python
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:78]:
```suggestion
A pipeline that implements a simple ETL, that has steps for branch creation and commits.
    
   ```python
```

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: docs/integrations/kubeflow.md (hunk 12 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md]: removed plural but kept the backslash to start a new line after "Note"

File: docs/integrations/kubeflow.md (hunk 18 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:40]: I'm losing numbering otherwise.. any other idea?

File: docs/integrations/kubeflow.md (hunk 19 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:76]: Yes, I need it because I'm trying to exclude this section from the toc.  Am I missing anything?

File: docs/integrations/kubeflow.md (hunk 8 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:95]: good point

File: docs/integrations/kubeflow.md (hunk 13 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md]: 1. Changed it, thanks. 
2. I changed the text so this became irrelevant

File: docs/integrations/kubeflow.md (hunk 11 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:90]: Agreed, thanks

File: docs/integrations/kubeflow.md (hunk 19 above)
Comment (nopcoder) [reply docs/integrations/kubeflow.md:76]: my bad, missed the part that this page got toc

File: docs/integrations/kubeflow.md [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+def lakectl_pipeline():
+   create_branch_task = create_branch_op()
+   extract_task = example_extract_op() 
+   commit_task = commit_op() 
+   transform_task = example_transform_op()
+   commit_task = commit_op()
+   load_task = example_load_op()
+```
+
+**Note**  \
```
Comment (nopcoder) [inline docs/integrations/kubeflow.md:89]: why do you force line break using '\' - from over notes we have, we don't force split here

File: docs/integrations/kubeflow.md (hunk 2 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md]: Following @johnnyaug 's question, I changed the structure of the docs. can you give it another look and let me know your thoughts? 

@nopcoder 
1. I added a note mentioning this.. let me know if you think this is redundant
2. 100% agree, IMO we should do this when we develop an official proper Kubeflow integration. WDYT?

File: docs/integrations/kubeflow.md (hunk 18 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md]: Fixed it, thanks

**Review by nopcoder — APPROVED** [review]

File: docs/integrations/kubeflow.md (hunk 21 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md]: It looked nicer to me, but I changed it now to follow the convention.

File: docs/integrations/kubeflow.md [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+description: Easily build reproducible data pipelines with Kubeflow and lakeFS using commits, without modifying the code or logic of your job. 
+parent: Integrations
+nav_order: 56
+has_children: false
+
+---
+# Using lakeFS with Kubeflow pipelines
+{: .no_toc }
+[Kubeflow](https://www.kubeflow.org/docs/about/kubeflow/) is a project dedicated to making deployments of ML workflows on Kubernetes simple, portable and scalable.
+A Kubeflow pipeline is a portable and scalable definition of an ML workflow composed of steps. Each step in the pipeline is an instance of a component represented as an instance of [ContainerOp](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ContainerOp).
```
Comment (johnnyaug) [inline docs/integrations/kubeflow.md:13]:
```suggestion
A Kubeflow pipeline is a portable and scalable definition of an ML workflow composed of steps. Each step in the pipeline is represented as an instance of [ContainerOp](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ContainerOp).
```
Or am I wrong

**Review by johnnyaug — APPROVED** [review]

File: docs/_config.yml (hunk 1 above)
Comment (talSofer) [reply docs/_config.yml]: I removed this. I had a config file under doc/bundle directory that made the vendor directory get generated under the docs pkg. I deleted the bundle and vendor dirs form my local and now make docs-serve works without excluding "vendor".

File: docs/integrations/kubeflow.md (hunk 22 above)
Comment (talSofer) [reply docs/integrations/kubeflow.md:13]: as discussed f2f , I'm using the Kubeflow documentation wording so we will stick to it.
